txnsync/bitmask.go

```diff
@@ -0,0 +1,226 @@
+			numExists++
+		} else {
+			lastNotExists = i
+		}
+	}
+	bitmaskType := 0
+	bestSize := bytesNeededBitmask(lastExists)
+	if bestSize > bytesNeededBitmask(lastNotExists) {
+		bitmaskType = 1
+		bestSize = bytesNeededBitmask(lastNotExists)
+	}
+	if bestSize > numExists*2+1 {
+		bitmaskType = 2
+		bestSize = numExists*2 + 1
+	}
+	if bestSize > (entries-numExists)*2+1 {
+		bitmaskType = 3
+		bestSize = (entries-numExists)*2 + 1
+	}
+	switch bitmaskType {
+	case 1:
+		(*b)[0] = 1
+		for i := range *b {
+			if i != 0 {
+				(*b)[i] = 255 - (*b)[i] // invert bits
+			}
+		}
+	case 2:
+		newBitmask := make(bitmask, 1, bestSize)
+		newBitmask[0] = 2
+		last := 0
+		for i := 0; i < entries; i++ {
+			byteIndex := i/8 + 1
+			if (*b)[byteIndex]&(1<<(i%8)) != 0 {
+				diff := i - last
+				newBitmask = append(newBitmask, byte(diff/256), byte(diff%256))
+				last = i
+			}
+		}
+		*b = newBitmask
+		return
+	case 3:
+		newBitmask := make(bitmask, 1, bestSize)
+		newBitmask[0] = 3
+		last := 0
+		for i := 0; i < entries; i++ {
+			byteIndex := i/8 + 1
+			if (*b)[byteIndex]&(1<<(i%8)) == 0 {
+				diff := i - last
+				newBitmask = append(newBitmask, byte(diff/256), byte(diff%256))
+				last = i
+			}
+		}
+		*b = newBitmask
+		return
+	default:
+	}
+
+	*b = bytes.TrimRight(*b, string(0))
+}
+
+func (b *bitmask) expandBitmask(entries int) {
+	option := 0
+	if len(*b) > 0 {
+		option = int((*b)[0])
+	} else {
+		return
+	}
+	switch option {
+	case 0: // if we have the bit 1 then we have an entry at the corresponding bit index.
+		return
+	case 1: // if we have the bit 0 then we have an entry at the corresponding bit index.
+		newBitmask := make(bitmask, bytesNeededBitmask(entries))
+		for i := range newBitmask {
+			if i != 0 {
+				if i < len(*b) {
+					newBitmask[i] = 255 - (*b)[i] // invert bits
+				} else {
+					newBitmask[i] = 255
+				}
+			}
+		}
+		*b = newBitmask
+	case 2: // contains a list of bytes designating the transaction bit index
+		newBitmask := make(bitmask, bytesNeededBitmask(entries))
+		sum := 0
+		for i := 0; i*2+2 < len(*b); i++ {
+			sum += int((*b)[i*2+1])*256 + int((*b)[i*2+2])
+			newBitmask.SetBit(sum)
+		}
+		*b = newBitmask
+	case 3: // contain a list of bytes designating the negative transaction bit index
+		newBitmask := make(bitmask, bytesNeededBitmask(entries))
+		sum := 0
+		for i := 0; i*2+2 < len(*b); i++ {
+			sum += int((*b)[i*2+1])*256 + int((*b)[i*2+2])
+			newBitmask.SetBit(sum)
+		}
+		*b = newBitmask
+		for i := range *b {
+			if i != 0 {
+				(*b)[i] = 255 - (*b)[i] // invert bits
+			}
+		}
+	}
+}
+
+func (b *bitmask) Iterate(entries int, callback func(i int) error) error {
+	option := 0
+	if len(*b) > 0 {
+		option = int((*b)[0])
+	} else { // nothing to iterate
+		return nil
+	}
+	switch option {
+	case 0:
+		for i := 0; i < entries; i++ {
+			byteIndex := i/8 + 1
+			if byteIndex < len(*b) && ((*b)[byteIndex]&(1<<(i%8)) != 0) {
+				if err := callback(i); err != nil {
+					return err
+				}
+			}
+		}
```

Comment:
hmm. I don't think you want to iterate on entries, as you'd have the same inefficiency for which you implemented this method to avoid. instead,
```golang
for i, v := range b[1:] {
   for j := 0; j < 8 && v > 0; j++, v>>=1 {
       if v & 1 != 0 {
            if err := callback(8*i+j); err != nil {
                 return err
            }
       }
   }
}
```